File: apps/echo-satellite/src-tauri/src/spy_links.rs
use crate::config::SpyMirageLink;
use reqwest::Client;
use serde::Deserialize;
use serde_json::Value;
use std::time::Duration;

#[derive(Debug, Clone, Default)]
pub struct SpyLinksSnapshot {
    pub reachable: bool,
    pub mirages: Vec<SpyMirageLink>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct SpyCodesResponse {
    ok: Option<bool>,
    paired_mirages: Option<Vec<SpyMirageLink>>,
    paired_mirage: Option<SpyMirageLink>,
}

fn candidate_ports() -> Vec<u16> {
    let mut ports = vec![3000, 3001, 8080];
    if let Ok(raw) = std::env::var("ECHO_CYBERDECK_HTTP_PORT") {
        for part in raw.split(',') {
            if let Ok(port) = part.trim().parse::<u16>() {
                if port > 0 && !ports.contains(&port) {
                    ports.push(port);
                }
            }
        }
    }
    ports
}

fn extract_mirages(payload: &SpyCodesResponse) -> Vec<SpyMirageLink> {
    if let Some(mirages) = payload.paired_mirages.as_ref() {
        if !mirages.is_empty() {
            return mirages.clone();
        }
    }
    payload
        .paired_mirage
        .clone()
        .map(|mirage| vec![mirage])
        .unwrap_or_default()
}

fn extract_mirages_from_value(value: &Value) -> Vec<SpyMirageLink> {
    if let Ok(payload) = serde_json::from_value::<SpyCodesResponse>(value.clone()) {
        return extract_mirages(&payload);
    }
    Vec::new()
}

async fn read_state_file_mirages() -> Option<Vec<SpyMirageLink>> {
    let state_path = std::env::var("ECHO_MIRAGE_SPY_PAIRING_STATE_PATH")
        .ok()
        .filter(|value| !value.trim().is_empty())
        .unwrap_or_else(|| ".tmp/echo-spy-pairing.json".to_string());

    let raw = tokio::fs::read_to_string(state_path).await.ok()?;
    let value: Value = serde_json::from_str(&raw).ok()?;
    let mirages = extract_mirages_from_value(&value);
    Some(mirages)
}

pub async fn fetch_spy_mirage_links(client: &Client) -> SpyLinksSnapshot {
    for port in candidate_ports() {
        let url = format!("http://127.0.0.1:{port}/api/spy/echo/codes");
        let Ok(response) = client
            .get(url)
            .timeout(Duration::from_secs(2))
            .send()
            .await
        else {
            continue;
        };
        if !response.status().is_success() {
            continue;
        }
        let Ok(payload) = response.json::<SpyCodesResponse>().await else {
            continue;
        };
        if payload.ok.unwrap_or(false) {
            return SpyLinksSnapshot {
                reachable: true,
                mirages: extract_mirages(&payload),
            };
        }
    }

    if let Some(mirages) = read_state_file_mirages().await {
        return SpyLinksSnapshot {
            reachable: true,
            mirages,
        };
    }

    SpyLinksSnapshot::default()
}

File: apps/echo-satellite-electron/electron/spy-links.mjs
import fs from "node:fs/promises";
import path from "node:path";

const DEFAULT_PORTS = [3000, 3001, 8080];

/** @typedef {{ nodeId: string, pairedAt: string }} SpyMirageLink */

/** @returns {number[]} */
function candidatePorts() {
  const fromEnv = process.env.ECHO_CYBERDECK_HTTP_PORT?.trim();
  const ports = new Set(DEFAULT_PORTS);
  if (fromEnv) {
    for (const part of fromEnv.split(",")) {
      const port = Number(part.trim());
      if (Number.isFinite(port) && port > 0) {
        ports.add(port);
      }
    }
  }
  return [...ports];
}

/** @param {unknown} value @returns {value is SpyMirageLink} */
function isMirageLink(value) {
  if (!value || typeof value !== "object") return false;
  const record = /** @type {Record<string, unknown>} */ (value);
  return typeof record.nodeId === "string" && typeof record.pairedAt === "string";
}

/** @param {unknown} payload @returns {SpyMirageLink[]} */
function extractMirages(payload) {
  if (!payload || typeof payload !== "object") return [];
  const data = /** @type {Record<string, unknown>} */ (payload);
  if (Array.isArray(data.pairedMirages)) {
    return data.pairedMirages.filter(isMirageLink);
  }
  if (isMirageLink(data.pairedMirage)) {
    return [data.pairedMirage];
  }
  return [];
}

/** @returns {Promise<SpyMirageLink[] | null>} */
async function readStateFileMirages() {
  const statePath =
    process.env.ECHO_MIRAGE_SPY_PAIRING_STATE_PATH?.trim() ||
    path.join(process.cwd(), ".tmp", "echo-spy-pairing.json");
  try {
    const raw = JSON.parse(await fs.readFile(statePath, "utf8"));
    return extractMirages(raw);
  } catch {
    return null;
  }
}

/**
 * Poll local cyberdeck Spy pairing state for linked Mirages.
 * @returns {Promise<{ reachable: boolean, mirages: SpyMirageLink[] }>}
 */
export async function fetchSpyMirageLinks() {
  for (const port of candidatePorts()) {
    try {
      const res = await fetch(`http://127.0.0.1:${port}/api/survey/echo/codes`, {
        signal: AbortSignal.timeout(2000),
      });
      if (!res.ok) continue;
      const data = await res.json();
      if (data?.ok) {
        return { reachable: true, mirages: extractMirages(data) };
      }
    } catch {
      /* try next port */
    }
  }

  const fromFile = await readStateFileMirages();
  if (fromFile) {
    return { reachable: true, mirages: fromFile };
  }

  return { reachable: false, mirages: [] };
}
